File: src/main/java/org/taktik/icure/services/external/rest/v1/dto/DocumentDto.java
/*
 * Copyright (C) 2018 Taktik SA
 *
 * This file is part of iCureBackend.
 *
 * iCureBackend is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as published by
 * the Free Software Foundation.
 *
 * iCureBackend is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with iCureBackend.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.taktik.icure.services.external.rest.v1.dto;

import org.taktik.icure.services.external.rest.v1.dto.embed.DocumentLocation;
import org.taktik.icure.services.external.rest.v1.dto.embed.DocumentStatus;
import org.taktik.icure.services.external.rest.v1.dto.embed.DocumentType;
import org.taktik.icure.services.external.rest.v1.dto.embed.DelegationDto;

import java.util.*;

@SuppressWarnings("UnusedDeclaration")
public class DocumentDto extends IcureDto {
    protected byte[] attachment;
	protected String attachmentId;
	protected DocumentLocation documentLocation;
	protected DocumentType documentType;
	protected DocumentStatus documentStatus;
	protected String mainUti;
	protected String name;
	protected Set<String> otherUtis;

	protected String externalUri;

	protected String storedICureDocumentId;


	public String getAttachmentId() {
		return attachmentId;
	}

	public void setAttachmentId(String attachmentId) {
		this.attachmentId = attachmentId;
	}

    public byte[] getAttachment() {
        return attachment;
    }

    public void setAttachment(byte[] attachment) {
        this.attachment = attachment;
    }

    public DocumentType getDocumentType() {
		return documentType;
	}

	public void setDocumentType(DocumentType documentType) {
		this.documentType = documentType;
	}

	public String getName() {
		return name;
	}

	public void setName(String name) {
		this.name = name;
	}

	public String getMainUti() {
		return mainUti;
	}

	public void setMainUti(String mainUti) {
		this.mainUti = mainUti;
	}

	public Set<String> getOtherUtis() {
		return otherUtis;
	}

	public void setOtherUtis(Set<String> otherUtis) {
		this.otherUtis = otherUtis;
	}

	public String getExternalUri() {
		return externalUri;
	}

	public void setExternalUri(String externalUri) {
		this.externalUri = externalUri;
	}

	public String getStoredICureDocumentId() {
		return storedICureDocumentId;
	}

	public void setStoredICureDocumentId(String storedICureDocumentId) {
		this.storedICureDocumentId = storedICureDocumentId;
	}

	public DocumentStatus getDocumentStatus() {
		return documentStatus;
	}

	public void setDocumentStatus(DocumentStatus documentStatus) {
		this.documentStatus = documentStatus;
	}

	public DocumentLocation getDocumentLocation(){
		return documentLocation;
	}

	public void setDocumentLocation(DocumentLocation documentLocation){
		this.documentLocation = documentLocation;
	}

}

File: src/main/kotlin/org/taktik/icure/services/external/rest/v1/dto/DocumentDto.kt
/*
 *  iCure Data Stack. Copyright (c) 2020 Taktik SA
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU Affero General Public License as
 *     published by the Free Software Foundation, either version 3 of the
 *     License, or (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful, but
 *     WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *     Affero General Public License for more details.
 *
 *     You should have received a copy of the GNU Affero General Public
 *     License along with this program.  If not, see
 *     <https://www.gnu.org/licenses/>.
 */
package org.taktik.icure.services.external.rest.v1.dto


import com.fasterxml.jackson.annotation.JsonIgnoreProperties
import com.fasterxml.jackson.annotation.JsonInclude
import com.github.pozo.KotlinBuilder
import io.swagger.v3.oas.annotations.media.Schema
import org.taktik.icure.services.external.rest.v1.dto.base.CodeStubDto
import org.taktik.icure.services.external.rest.v1.dto.base.EncryptableDto
import org.taktik.icure.services.external.rest.v1.dto.base.ICureDocumentDto
import org.taktik.icure.services.external.rest.v1.dto.base.StoredDocumentDto
import org.taktik.icure.services.external.rest.v1.dto.embed.DelegationDto
import org.taktik.icure.services.external.rest.v1.dto.embed.DocumentLocationDto
import org.taktik.icure.services.external.rest.v1.dto.embed.DocumentStatusDto
import org.taktik.icure.services.external.rest.v1.dto.embed.DocumentTypeDto

@JsonInclude(JsonInclude.Include.NON_NULL)
@JsonIgnoreProperties(ignoreUnknown = true)
@KotlinBuilder
@Schema(description = """This entity is a root level object. It represents a Document. It is serialized in JSON and saved in the underlying CouchDB database.""")
data class DocumentDto(
        @Schema(description = "The Id of the document. We encourage using either a v4 UUID or a HL7 Id.") override val id: String,
        @Schema(description = "The revision of the document in the database, used for conflict management / optimistic locking.") override val rev: String? = null,
        override val created: Long? = null,
        override val modified: Long? = null,
        override val author: String? = null,
        override val responsible: String? = null,
        override val medicalLocationId: String? = null,
        override val tags: Set<CodeStubDto> = setOf(),
        override val codes: Set<CodeStubDto> = setOf(),
        override val endOfLife: Long? = null,
        override val deletionDate: Long? = null,

        @Schema(description = "Reference in object store") val objectStoreReference: String? = null,
        @Schema(description = "Location of the document") val documentLocation: DocumentLocationDto? = null,
        @Schema(description = "The type of document, ex: admission, clinical path, document report,invoice, etc.") val documentType: DocumentTypeDto? = null,
        @Schema(description = "The status of the development of the document. Ex: Draft, finalized, reviewed, signed, etc.") val documentStatus: DocumentStatusDto? = null,
        @Schema(description = "When the document is stored in an external repository, this is the uri of the document in that repository") val externalUri: String? = null,
        @Schema(description = "The main Uniform Type Identifier of the document (https://developer.apple.com/library/archive/documentation/FileManagement/Conceptual/understanding_utis/understand_utis_conc/understand_utis_conc.html#//apple_ref/doc/uid/TP40001319-CH202-CHDHIJDE)") val mainUti: String? = null,
        @Schema(description = "Name of the document") val name: String? = null,
        @Schema(description = "The document version") val version: String? = null,
        @Schema(description = "Extra Uniform Type Identifiers") val otherUtis: Set<String> = setOf(),
        @Schema(description = "The ICureDocument (Form, Contact, ...) that has been used to generate the document") val storedICureDocumentId: String? = null, //The ICureDocumentDto (FormDto, ContactDto, ...) that has been used to generate the document
        @Schema(description = "A unique external id (from another external source).") val externalUuid: String? = null,
        @Schema(description = "Size of the document file") val size: Long? = null,
        @Schema(description = "Hashed version of the document") val hash: String? = null,
        @Schema(description = "Id of the contact during which the document was created") val openingContactId: String? = null,

        @Schema(description = "Id of attachment to this document") val attachmentId: String? = null,

        val encryptedAttachment: ByteArray? = null,
        val decryptedAttachment: ByteArray? = null,

        override val secretForeignKeys: Set<String> = setOf(),
        override val cryptedForeignKeys: Map<String, Set<DelegationDto>> = mapOf(),
        override val delegations: Map<String, Set<DelegationDto>> = mapOf(),
        override val encryptionKeys: Map<String, Set<DelegationDto>> = mapOf(),

        override val encryptedSelf: String? = null
) : StoredDocumentDto, ICureDocumentDto<String>, EncryptableDto {
    override fun withIdRev(id: String?, rev: String) = if (id != null) this.copy(id = id, rev = rev) else this.copy(rev = rev)
    override fun withDeletionDate(deletionDate: Long?) = this.copy(deletionDate = deletionDate)
}
